src/peerconnection.cpp

```diff
@@ -283,6 +283,11 @@ void PeerConnection::onDataChannel(
 }
 
 std::shared_ptr<Track> PeerConnection::addTrack(Description::Media description) {
+	auto local = impl()->localDescription();
+	if (!impl()->config.forceTrackAllocation && !(local && local->hasAudioOrVideo())) {
+		throw std::runtime_error("addTrack() called, but srtp transport was not initialized. This is an optimization for use of the library with data channels only. Set config.forceTrackAllocation to 'true' to get the transport initialized before dynamically adding tracks.");
+	}
```

Comment:
You can't do such a check here as most of the time `addTrack()` is called before generating the local offer, so it will throw when `forceTrackAllocation` is false. That's why existing tests for tracks fail.

You could instead prevent opening the track, logging an error, and maybe calling `track->triggerError()` here: https://github.com/paullouisageneau/libdatachannel/blob/99be4b7451fc68a9d52dc307b766c503fb153f8f/src/impl/peerconnection.cpp#L759

Reply:
Something like this?

Reply:
Yes it looks good!